examples/contrib/forecast/bart.py

```diff
@@ -0,0 +1,149 @@
+
+
+# We define a model by subclassing the ForecastingModel class and implementing
+# a single .model() method.
+class Model(ForecastingModel):
+    # The .model() method inputs two tensors: a fake tensor zero_data that is
+    # the same size and dtype as the real data (but of course the generative
+    # model shouldn't depend on the value of the data it generates!), and a
+    # tensor of covariates. Our simple model depends on no covariates, so we
+    # simply pass in an empty tensor (see  the preprocess() function above).
+    def model(self, zero_data, covariates):
+        duration = zero_data.size(-2)
+        period = 24 * 7
+
+        # Sample global parameters.
+        noise_scale = pyro.sample("noise_scale",
+                                  dist.LogNormal(torch.full((2,), -3), 1).to_event(1))
+        trans_timescale = pyro.sample("trans_timescale",
+                                      dist.LogNormal(torch.zeros(2), 1).to_event(1))
+
+        trans_scale = pyro.sample("trans_scale",
+                                  dist.LogNormal(torch.zeros(2), 0.1).to_event(1))
+        trans_corr = pyro.sample("trans_corr",
+                                 dist.LKJCorrCholesky(2, torch.ones(())))
+        trans_scale_tril = trans_scale.unsqueeze(-1) * trans_corr
+
+        obs_scale = pyro.sample("obs_scale",
+                                dist.LogNormal(torch.zeros(2), 0.1).to_event(1))
+        obs_corr = pyro.sample("obs_corr",
+                               dist.LKJCorrCholesky(2, torch.ones(())))
+        obs_scale_tril = obs_scale.unsqueeze(-1) * obs_corr
+
+        # Note the initial seasonality should be sampled in a plate with the
+        # same dim as the time_plate, dim=-1. That way we can repeat the dim
+        # below using periodic_repeat().
+        with pyro.plate("season_plate", period,  dim=-1):
+            season_init = pyro.sample("season_init",
+                                      dist.Normal(torch.zeros(2), 1).to_event(1))
+
+        # Sample independent noise at each time step.
+        with self.time_plate:
+            season_noise = pyro.sample("season_noise",
+                                       dist.Normal(0, noise_scale).to_event(1))
+
+        # Construct a prediction. This prediction has an exactly repeated
+        # seasonal part plus slow seasonal drift. We use two deterministic,
+        # linear functions to transform our diagonal Normal noise to nontrivial
+        # samples from a Gaussian process.
+        prediction = (periodic_repeat(season_init, duration, dim=-2) +
+                      periodic_cumsum(season_noise, period, dim=-2))
+
+        # Construct a joint noise model. This model is a GaussianHMM, whose
+        # .rsample() and .log_prob() methods are parallelized over time; this
+        # this entire model is parallelized over time.
+        init_dist = dist.Normal(torch.zeros(2), 100).to_event(1)
+        trans_mat = trans_timescale.neg().exp().diag_embed()
```

Comment:
ou process?

Reply:
Yes, this is an OU process around the seasonal GP